services/paymaster/src/constraint/tolls.rs

```diff
@@ -0,0 +1,100 @@
+use crate::rpc::ChainIndex;
+use borsh::BorshDeserialize;
+use fogo_sessions_sdk::tollbooth::TOLLBOOTH_PROGRAM_ID;
+use reqwest::StatusCode;
+use serde::Deserialize;
+use serde_with::{serde_as, DisplayFromStr};
+use solana_program::instruction::CompiledInstruction;
+use solana_pubkey::Pubkey;
+use solana_transaction::versioned::VersionedTransaction;
+
+#[derive(Default, Deserialize, Clone)]
+#[serde_as]
+pub enum Tolls {
+    #[serde(rename = "free")]
+    #[default]
+    Free,
+    #[serde(rename = "fixed")]
+    Fixed(Vec<Toll>),
+}
+
+#[serde_as]
+#[derive(Deserialize, Clone)]
+
+pub struct Toll {
+    amount: u64,
+    #[serde_as(as = "DisplayFromStr")]
+    mint: Pubkey,
+}
+
+impl Tolls {
+    const MINT_ACCOUNT_INDEX_PAY_TOLL_INSTRUCTION: usize = 4;
+
+    pub async fn validate_toll_payment(
+        &self,
+        transaction: &VersionedTransaction,
+        chain_index: &ChainIndex,
+    ) -> Result<(), (StatusCode, String)> {
+        let tollbooth_instructions: Vec<(usize, &CompiledInstruction)> = transaction
+            .message
+            .instructions()
+            .iter()
+            .enumerate()
+            .filter(|(_, ix)| {
+                ix.program_id(transaction.message.static_account_keys()) == &TOLLBOOTH_PROGRAM_ID
+            })
+            .collect();
+
+        match (self, tollbooth_instructions.as_slice()) {
+            (Tolls::Free, []) => Ok(()),
+            (Tolls::Free, _) => Err((
+                StatusCode::BAD_REQUEST,
+                "The paymaster for this domain does not require a toll payment".to_string(),
+            )),
```

Comment:
I don't think this should error; it does not hurt for the toll to be sent even when it isn't strictly required and it makes rollouts of changes easier, because it removes the requirement that the frontend change is synced to the paymaster config change.

Even if the paymaster were to e.g. expose an api that indicates current tolls, there's still a race condition where a tx created before the paymaster config was changed would fail if landed after the paymaster config changed.

IMO as long as the tx pays the requested toll _or more_, there shouldn't be a reason to stop the tx from landing.